File: Leetcode/Python/_202.py
class Solution:
    def isHappy(self, num: int) -> bool:
        def decompose(num):
            output = 0
          
            while num:
                output += (num % 10) ** 2
                num = num // 10
            return output
        
        slow = num
        fast = num
        
        while True:
            slow = decompose(slow)
            fast = decompose(decompose(fast))
            if slow == fast:
                break
            
        return slow == 1

File: Leetcode/cpp/202.cpp
class Solution {
public:
    bool isHappy(int n) {
        std::set<int> seen;

        while (n > 1) {
            int tmp = 0;
            while (n) {
                tmp += (n % 10) * (n % 10);
                n /= 10;
            }
            if (seen.contains(tmp)) {
                return false;
            }
            seen.insert(tmp);
            n = tmp;
        }
        return n == 1;
    }
};
